src/System.Linq/src/System/Linq/SequenceEqual.cs

```diff
@@ -28,9 +28,25 @@ public static bool SequenceEqual<TSource>(this IEnumerable<TSource> first, IEnum
                 throw Error.ArgumentNull(nameof(second));
             }
 
-            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol && firstCol.Count != secondCol.Count)
+            if (first is ICollection<TSource> firstCol && second is ICollection<TSource> secondCol)
             {
-                return false;
+                if (firstCol.Count != secondCol.Count)
+                {
+                    return false;
+                }
+
+                if (firstCol is IReadOnlyList<TSource> firstList && secondCol is IReadOnlyList<TSource> secondList)
```

Comment:
`ImmutableList` doesn't implement `IList` so you won't have the performance penalty for it that you would have if you used `IReadOnlyList` as it will just use the regular `IEnumerable` path